rasterio/warp.py

```diff
@@ -241,6 +244,14 @@ def reproject(
         **kwargs)
 
 
+def res_tuple(res):
+    try:
```

Comment:
Duck typing and "Ask for forgiveness, not permission" tend to be the python way of handling these case. But if it feels funky, we could just test for it's iterable-ness.

>>> import collections
>>> isinstance([12, 13], collections.Iterable)
True

In terms of inlining the logic, I think I'd prefer to keep it separate. Big picture, I think we should aim to separate out the argument handling from our algorithms. It tends to make both easier to test and understand.